### PR TITLE
#17: Create buckets

### LINKED ISSUE #17: Creating new buckets
## Problem
As a user, I'd like to be able to create new buckets via this library

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #17

### CODE REVIEW

File: src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Executes {@link RpcCommand}s by converting them to a request, sending the request and processing the result.
+ */
+class JsonRpcCommandExecutor {
+    private final JsonRpcClient client;
+    private final JsonMapper jsonMapper;
+
+    JsonRpcCommandExecutor(final JsonRpcClient client, final JsonMapper jsonMapper) {
+        this.client = client;
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java:14]: You used `Objects.requireNotNull()` in one place, but not here. Why?

File: src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    <R> R execute(final RpcCommand<R> command) {
+        final Object parameters = command.getParameters();
+        final JsonObject jsonObjectParam = this.jsonMapper.toJsonObject(parameters);
+        final String responseBody = this.client
+                .sendRequest(new JsonRpcPayload("job_exec", command.getJobName(), jsonObjectParam));
+        return command.processResult(responseBody);
+    }
+
+    // Must be public for json mapping
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java:26]:
```suggestion
    // Must be public for JSON mapping
```

File: src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        public String getJob() {
+            return this.job;
+        }
+
+        public Parameters getParameters() {
+            return this.parameters;
+        }
+    }
+
+    // Must be public for json mapping
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/jsonrpc/JsonRpcCommandExecutor.java:54]:
```suggestion
    // Must be public for JSON mapping
```

File: src/main/java/com/exasol/bucketfs/jsonrpc/RpcCommand.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package com.exasol.bucketfs.jsonrpc;
+
+/**
+ * A command with job name and parameters that can be executed against the RPC interface. Child classes can process the
+ * response payload and return a custom model by implementing {@link #processResult(String)}.
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/jsonrpc/RpcCommand.java:5]: Explain "custom model" in this context, please.

Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/jsonrpc/RpcCommand.java:5]: Explain "job".

File: src/main/java/com/exasol/bucketfs/jsonrpc/RpcCommand.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+     * The job name.
+     *
+     * @return the job name
+     */
+    final String getJobName() {
+        return this.jobName;
+    }
+
+    /**
+     * Parameters for the RPC request. The returned object must be serializable to Json.
```
Comment (redcatbear) [inline src/main/java/com/exasol/bucketfs/jsonrpc/RpcCommand.java:31]:
```suggestion
     * Parameters for the RPC request.
     * <p>
     * The returned object must be serializable to JSON.
     * </p>
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/BasicAuthAuthenticatorTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+class BasicAuthAuthenticatorTest {
+
+    @Test
+    void testAuthenticate() throws URISyntaxException {
+        final var requestBuilder = HttpRequest.newBuilder(new URI("http://localhost"));
+
+        new BasicAuthAuthenticator("user", "password").authenticate(requestBuilder);
+
+        assertAuthHeader(requestBuilder, "Basic dXNlcjpwYXNzd29yZA==");
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/BasicAuthAuthenticatorTest.java:21]:
```suggestion
        final var requestBuilder = HttpRequest.newBuilder(new URI("http://localhost"));
        new BasicAuthAuthenticator("user", "password").authenticate(requestBuilder);
        assertAuthHeader(requestBuilder, "Basic dXNlcjpwYXNzd29yZA==");
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/BearerTokenAuthenticatorTest.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+class BearerTokenAuthenticatorTest {
+
+    @Test
+    void testAuthenticate() throws URISyntaxException {
+        final var requestBuilder = HttpRequest.newBuilder(new URI("http://localhost"));
+
+        new BearerTokenAuthenticator("token").authenticate(requestBuilder);
+
+        assertAuthHeader(requestBuilder, "Bearer token");
+    }
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/BearerTokenAuthenticatorTest.java:22]:
```suggestion
    void testAuthenticate() throws URISyntaxException {
        final var requestBuilder = HttpRequest.newBuilder(new URI("http://localhost"));
        new BearerTokenAuthenticator("token").authenticate(requestBuilder);
        assertAuthHeader(requestBuilder, "Bearer token");
    }
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/BearerTokenAuthenticatorTest.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+    private void assertAuthHeader(final Builder requestBuilder, final String expectedAuthHeader) {
+        final var request = requestBuilder.build();
+        final var authHeaders = request.headers().allValues("Authorization");
+        assertThat(authHeaders, contains(expectedAuthHeader));
+    }
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/BearerTokenAuthenticatorTest.java:28]: Code duplication. I suggest a static-only assertion class.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+import com.exasol.bucketfs.*;
+import com.exasol.bucketfs.jsonrpc.CommandFactory.Builder;
+import com.exasol.bucketfs.jsonrpc.CreateBucketCommand.CreateBucketCommandBuilder;
+import com.exasol.containers.ExasolDockerImageReference;
+
+@Tag("slow")
+// [itest->dsn~creating-new-bucket~1]
+class CreateBucketCommandIT extends AbstractBucketIT {
+
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:24]:
```suggestion
class CreateBucketCommandIT extends AbstractBucketIT {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    @Test
+    void creatingBucketWithCheckingCertificateFails() throws BucketAccessException, TimeoutException {
+        assumeJsonRpcAvailable();
+
+        final String randomBucketName = getRandomBucketName();
+
+        final CreateBucketCommandBuilder command = createCommandFactory(false).makeCreateBucketCommand()
+                .bucketFsName(DEFAULT_BUCKETFS).bucketName(randomBucketName).isPublic(true).readPassword(READ_PASSWORD)
+                .writePassword(WRITE_PASSWORD);
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:36]: One line per chain link please.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+        final JsonRpcException exception = assertThrows(JsonRpcException.class, () -> command.execute());
+        assertThat(exception.getMessage(), containsString("Error executing request"));
+        assertThat(exception.getCause().getMessage(), equalTo(
+                "PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target"));
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:41]: Please wrap in `assertAll`.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+@Tag("slow")
+// [itest->dsn~creating-new-bucket~1]
+class CreateBucketCommandIT extends AbstractBucketIT {
+
+    private static final String READ_PASSWORD = "READ_PASSWORD";
+    private static final String WRITE_PASSWORD = "WRITE_PASSWORD";
+
+    @Test
+    void creatingBucketWithCheckingCertificateFails() throws BucketAccessException, TimeoutException {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:29]:
```suggestion
    void testCreatingBucketWithCheckingCertificateThrowsException() throws BucketAccessException, TimeoutException {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                .writePassword(WRITE_PASSWORD);
+
+        final JsonRpcException exception = assertThrows(JsonRpcException.class, () -> command.execute());
+        assertThat(exception.getMessage(), containsString("Error executing request"));
+        assertThat(exception.getCause().getMessage(), equalTo(
+                "PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target"));
+    }
+
+    @Test
+    void creatingBucketWithExistingNameFails() throws BucketAccessException, TimeoutException {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:45]:
```suggestion
    void testCreatingBucketWithExistingNameThrowsException() throws BucketAccessException, TimeoutException {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+        command.execute();
+
+        final JsonRpcException exception = assertThrows(JsonRpcException.class, () -> command.execute());
+        assertThat(exception.getMessage(),
+                containsString("Given bucket " + randomBucketName + " already exists in bucketfs " + DEFAULT_BUCKETFS));
+    }
+
+    @Test
+    void createdBucketIsWriteable() throws BucketAccessException, TimeoutException, InterruptedException {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:65]:
```suggestion
    void testCreatedBucketIsWriteable() throws BucketAccessException, TimeoutException, InterruptedException {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+                .bucketName(randomBucketName) //
+                .isPublic(true) //
+                .readPassword(READ_PASSWORD) //
+                .writePassword(WRITE_PASSWORD) //
+                .execute();
+        assertBucketWritable(randomBucketName);
+    }
+
+    @Test
+    void createdBucketWithDefaultValuesSucceeds() throws BucketAccessException, TimeoutException, InterruptedException {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:81]:
```suggestion
    void testCreateBucketWithDefaultValues() throws BucketAccessException, TimeoutException, InterruptedException {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        assertDoesNotThrow(() -> createCommandFactory().makeCreateBucketCommand() //
+                .bucketFsName(DEFAULT_BUCKETFS) //
+                .bucketName(randomBucketName) //
+                .execute());
+    }
+
+    private void assumeJsonRpcAvailable() {
+        final ExasolDockerImageReference version = EXASOL.getDockerImageReference();
+        assumeTrue(version.getMajor() >= 7,
+                "JSON RPC only available with version >= 7, " + version + " is not supported.");
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:95]:
```suggestion
                "JSON RPC only available with Exasol version 7 or later, " + version + " is not supported.");
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    private String getRandomBucketName() {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:98]:
```suggestion
    private String getUniqueBucketName() {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+import org.junit.jupiter.api.Test;
+
+import com.exasol.bucketfs.*;
+import com.exasol.bucketfs.jsonrpc.CommandFactory.Builder;
+import com.exasol.bucketfs.jsonrpc.CreateBucketCommand.CreateBucketCommandBuilder;
+import com.exasol.containers.ExasolDockerImageReference;
+
+@Tag("slow")
+// [itest->dsn~creating-new-bucket~1]
+class CreateBucketCommandIT extends AbstractBucketIT {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:23]: Please reorder methods in order of first reference.

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+    private void assertBucketWritable(final String randomBucketName)
+            throws BucketAccessException, TimeoutException, InterruptedException {
+        final SyncAwareBucket bucket = createBucket(randomBucketName);
+
+        waitUntilBucketExists(bucket);
+
+        final var fileName = "test-uploaded.txt";
+        bucket.uploadStringContent("file content", fileName);
+        assertThat(bucket.listContents(), hasItem(fileName));
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:125]: Nice. :+1:

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+
+    private void waitUntilBucketExists(final SyncAwareBucket bucket) throws InterruptedException {
+        final Instant start = Instant.now();
+        final Duration timeout = Duration.ofSeconds(5);
+
+        while (!bucketExists(bucket)) {
+            delayNextCheck();
+            final Duration waitingTime = Duration.between(start, Instant.now());
+            if (timeout.minus(waitingTime).isNegative()) {
+                fail("Bucket " + bucket.getBucketName() + " not available after " + waitingTime);
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:136]:
```suggestion
                fail("Time out trying to verify that Bucket " + bucket.getBucketName() + " exists after waiting " + waitingTime);
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+    @SuppressWarnings("java:S2925") // Sleep required for waiting until bucket is available
+    private void delayNextCheck() throws InterruptedException {
+        Thread.sleep(100);
+    }
+
+    private boolean bucketExists(final SyncAwareBucket bucket) {
+        try {
+            bucket.listContents();
+        } catch (final BucketAccessException e) {
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandIT.java:149]:
```suggestion
        } catch (final BucketAccessException exception) {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandTest.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import com.exasol.bucketfs.jsonrpc.CreateBucketCommand.CreateBucketCommandBuilder;
+import com.exasol.bucketfs.jsonrpc.CreateBucketCommand.Request;
+
+@ExtendWith(MockitoExtension.class)
+class CreateBucketCommandTest {
+
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandTest.java:20]:
```suggestion
class CreateBucketCommandTest {
```

File: src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    private Request getRpcRequest() {
+        final ArgumentCaptor<CreateBucketCommand> arg = ArgumentCaptor.forClass(CreateBucketCommand.class);
+        verify(this.commandExcecutorMock).execute(arg.capture());
+        assertThat(arg.getValue().getJobName(), equalTo("bucket_add"));
+        return arg.getValue().getParameters();
+    }
+
+    private void asserExecutionFails(final CreateBucketCommandBuilder builder,
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/jsonrpc/CreateBucketCommandTest.java:68]:
```suggestion
    private void assertCreateBucketCommandThrowsExceptionWithMessage(final CreateBucketCommandBuilder builder,
```

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: .github/workflows/ci-build.yml [hunk 24]
```diff
@@ -30,9 +33,34 @@ jobs:
           key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             ${{ runner.os }}-maven-
+      - name: Cache SonarQube packages
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply .github/workflows/ci-build.yml:36]: Already enabled a few lines before

File: .github/workflows/ci-build.yml [hunk 25]
```diff
@@ -30,9 +33,34 @@ jobs:
...
+              -Dsonar.organization=exasol -Dsonar.host.url=https://sonarcloud.io -Dsonar.login=$SONAR_TOKEN
         env:
-          GITHUB_OAUTH: ${{ secrets.GITHUB_TOKEN }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
\ No newline at end of file
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+      - name: Publish support package for ${{ matrix.docker_db_version }}
+        uses: actions/upload-artifact@v2
+        if: ${{ always() }}
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply .github/workflows/ci-build.yml:58]: No. When there are build errors (e.g. failed tests), the following steps won't be executed.

**Review by redcatbear — DISMISSED** [review]

**Review by redcatbear — APPROVED** [review]

**Pull request merged**